File: leetcode-go/094.Binary-Tree-Inorder-Traversal/inorder_traversal.go
package main

import "github.com/onwl007/leetcode-algo/structures"

type TreeNode = structures.TreeNode

/*
 * @lc app=leetcode.cn id=94 lang=golang
 *
 * [94] 二叉树的中序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func inorderTraversal(root *TreeNode) []int {
	vals := []int{}
	dfs(root, &vals)
	return vals
}

func dfs(root *TreeNode, vals *[]int) {
	if root == nil {
		return
	}
	dfs(root.Left, vals)
	*vals = append(*vals, root.Val)
	dfs(root.Right, vals)
}

// @lc code=end

func inorderIteration(root *TreeNode) (vals []int) {
	stack := []*TreeNode{}
	for len(stack) > 0 || root != nil {
		for root != nil {
			stack = append(stack, root)
			root = root.Left
		}
		root = stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		vals = append(vals, root.Val)
		root = root.Right
	}
	return
}

File: leetcode-go/094.Binary-Tree-Inorder-Traversal/inorder_traversal.js
/*
 * @lc app=leetcode.cn id=94 lang=javascript
 *
 * [94] 二叉树的中序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var inorderTraversal = function (root) {
  let res = [];
  dfs(root, res);
  return res;
};

function dfs(root, res) {
  if (!root) {
    return;
  }
  dfs(root.left, res);
  res.push(root.val);
  dfs(root.right, res);
}
// @lc code=end

var inorderTraversal1 = function (root) {
  const stack = [];
  const vals = [];
  while (stack.length > 0 || root) {
    while (root) {
      stack.push(root);
      root = root.left;
    }
    root = stack.pop();
    vals.push(root.val);
    root = root.right;
  }
  return vals;
};
